std/json.c3

```diff
@@ -0,0 +1,428 @@
+    allocator : Allocator,  // FIXME use global allocator.
+}
+
+typedef JSON_Node : {
+    type     : int, // 0: null, 1: true, 2: false, 3: string, 4: number,
+                    // 5: array, 6: object
+    value    : Optional<StringView>,
+    children : Optional<Vector<JSON_Node>>,
+}
+
+// TODO something like new<JSON_Node> would be nice, but we don't support
+// function specializations.
+function new_json_node : (node_type : int) -> JSON_Node = {
+    return {
+        node_type,
+        make_optional<StringView>(),
+        make_optional<Vector<JSON_Node>>(),
+    };
+}
+
+function char : (literal : u8[1]&) -> u8 = {
+    return literal[0];
+}
+
+function is_eof : (cursor : Cursor&) -> bool = {
+    return cursor.position >= cursor.data.length;
+}
+
+function peak : (cursor : Cursor&) -> Optional<u8> = {
+    if !cursor:is_eof() {
+        return make_optional(cursor.data.data[cursor.position]);
+    }
+
+    return make_optional<u8>();
+}
+
+function next : (cursor : Cursor&) -> Optional<u8> = {
+    // TODO this should be const (can't use with has_value() right now).
+    let token : Optional<u8> = cursor:peak();
+
+    if token:has_value() {
+        cursor.position += 1;
+    }
+
+    return token;
+}
+
+function is_json_whitespace : (token : u8) -> bool = {
+    // JSON defines whitespace as one of { space, linefeed, carriage return,
+    // horizontal tab }. If we use string literals instead of code points, then
+    // our code doesn't care about the underlying encoding :).
+    return token == char(" ") or token == char("\n")
+           or token == char("\r") or token == char("\t");
+}
+
+function eat : (cursor : Cursor&, token : u8) -> bool = {
+    // TODO this should be const (can't use with has_value() right now).
+    let next_char : Optional<u8> = cursor:peak();
+
+    let next_char_data : u8 = next_char:data();
+    if next_char:has_value() and next_char_data == token {
+        cursor:next();
+        return true;
+    }
+
+    return false;
+}
+
+function eat : (cursor : Cursor&, string : StringView) -> bool = {
+    for token in string:get_iter() {
+        if !cursor:eat(token) {
+            return false;
+        }
+    }
+
+    return true;
+}
+
+function eat_whitespace : (cursor: Cursor&) -> void = {
+    while !cursor:is_eof() {
+        let next_char : Optional<u8> = cursor:peak();
+        runtime_assert(next_char:has_value());
+
+        if !is_json_whitespace(next_char:data()) {
+            return;
+        }
+
+        // Eat the current character.
+        cursor:next();
+    }
+}
+
+// TODO return ErrorOr<JSON_Node>.
+function parse_object : (cursor : Cursor&) -> Optional<JSON_Node> = {
+    if !cursor:eat(char("{")) {
+        return make_optional<JSON_Node>();
+    }
+
+    let node : JSON_Node = new_json_node(/* object */ 6);
+
+    cursor:eat_whitespace();
+
+    if cursor:eat(char("}")) {
+        // Empty object.
+        return make_optional(node);
```

Comment:
NIT: imo an length 0 children vector is more ergonomic than a children vector optional since way you can do things like iterate through an object's children without worrying about whether it has children.

Lets also update the vector class to make the empty vector allocation free.

Reply:
I agree, but I didn't want the vectors to allocate every time you made a node.

> Lets also update the vector class to make the empty vector allocation free.

The question then is "what do we return when you call `data()` on an empty vector"? In C++, the standard says you may or may not get a null pointer (that you shouldn't dereference) either way. Are we happy to return an `Optional<Span<T>>`?

Reply:
In c++ it's hard because you have pointers. It's easy for us 🙃, you just return a zero length span -- now it's entirely the caller's fault for not checking.

The span's data pointer can just be undefined I don't think it matters, everywhere that takes a span should check there own bounds imo.

Reply:
I guess it's fine as long as you use the Span interface instead of accessing `.data`. Anw, I've changed it